crates/bevy_transform/src/systems.rs

```diff
@@ -89,100 +131,404 @@ pub fn propagate_transforms(
+            // SAFETY: The caller guarantees that `transform_query` will not be fetched for any
+            // descendants of `entity`, so it is safe to call `propagate_recursive` for each child.
+            //
+            // The above assertion ensures that each child has one and only one unique parent
+            // throughout the entire hierarchy.
+            unsafe {
+                propagate_recursive(
+                    global_matrix.as_ref(),
+                    transform_query,
+                    parent_query,
+                    child,
+                    changed || actual_parent.is_changed(),
+                );
+            }
+        }
+    }
+}
+
+/// Parallel hierarchy traversal with an optimized batching scheduler. Often 2-5 times faster than
+/// the serial version.
+///
+/// Relies on `std` until a `no_std` `mpsc` channel is available.
+#[cfg(feature = "std")]
+mod parallel {
+    use crate::prelude::*;
+    use bevy_ecs::{entity::UniqueEntityIter, prelude::*, system::lifetimeless::Read};
+    use bevy_tasks::{ComputeTaskPool, TaskPool};
+    use bevy_utils::Parallel;
+    use core::sync::atomic::{AtomicI32, Ordering};
+    use std::{
+        sync::{
+            // TODO: this implementation could be used in no_std if there was a
+            mpsc::{Receiver, Sender},
+            Arc,
+            Mutex,
+        },
+        vec::Vec,
+    };
+
+    /// Update [`GlobalTransform`] component of entities based on entity hierarchy and [`Transform`]
+    /// component.
+    ///
+    /// Third party plugins should ensure that this is used in concert with
+    /// [`sync_simple_transforms`](super::sync_simple_transforms) and
+    /// [`compute_transform_leaves`](super::compute_transform_leaves).
+    pub fn propagate_parent_transforms(
+        mut queue: Local<WorkQueue>,
+        mut orphaned: RemovedComponents<ChildOf>,
+        mut orphans: Local<Vec<Entity>>,
+        mut roots: Query<
+            (Entity, Ref<Transform>, &mut GlobalTransform, &Children),
+            Without<ChildOf>,
+        >,
+        nodes: NodeQuery,
+    ) {
+        // Orphans
+        orphans.clear();
+        orphans.extend(orphaned.read());
+        orphans.sort_unstable();
+
+        // Process roots in parallel, seeding the work queue
+        roots.par_iter_mut().for_each_init(
+            || queue.local_queue.borrow_local_mut(),
+            |outbox, (parent, transform, mut parent_transform, children)| {
+                if transform.is_changed()
+                    || parent_transform.is_added()
+                    || orphans.binary_search(&parent).is_ok()
+                {
+                    *parent_transform = GlobalTransform::from(*transform);
+                }
+
+                // SAFETY: the parent entities passed into this function are taken from iterating
+                // over the root entity query. Queries iterate over disjoint entities, preventing
+                // mutable aliasing, and making this call safe.
```

Comment:
Do we have any hierarchy validity enforcement such that it is not possible to have an entity be a child of more than one parent?

Reply:
This is done during propagation, to ensure there are no cycles when propagating. It would be nice if we could enforce this somehow, but with reflection, it's always possible to poison the hierarchy.

Reply:
i guess the consequences of such a graph, rather than a tree, would be just as nonsensical with the safe version?

Reply:
Yes, propagation would follow the same path, the only difference is that the unsafe one can be done in parallel. Note that our current impl is parallel at the roots only, is unsafe, and does this same test and panics if a cycle is found. The change here doesn't really change much in terms of unsafety - we are still pinky promising the compiler that our access to entities in parallel with unsafe is disjoint.